pkg/gitconfig/configs.go

```diff
@@ -0,0 +1,358 @@
+
+// SetLocal sets (or adds) a key only in the per-directory (local) config.
+func (c *Configs) SetLocal(key, value string) error {
+	if c.local == nil {
+		if c.workdir == "" {
+			return fmt.Errorf("no workdir set")
+		}
+		c.local = &Config{}
+	}
+
+	return c.local.Set(key, value)
+}
+
+// SetGlobal sets (or adds) a key only in the per-user (global) config.
+func (c *Configs) SetGlobal(key, value string) error {
+	if c.global == nil {
+		c.global = &Config{}
+	}
+
+	return c.global.Set(key, value)
+}
+
+// SetEnv sets (or adds) a key in the per-process (env) config. Useful
+// for persisting flags during the invocation.
+func (c *Configs) SetEnv(key, value string) error {
+	if c.env == nil {
+		c.env = &Config{
+			noWrites: true,
+		}
+	}
+
+	return c.env.Set(key, value)
+}
+
+// UnsetLocal deletes a key from the local config.
+func (c *Configs) UnsetLocal(key string) error {
+	if c.local == nil {
+		return nil
+	}
+
+	return c.local.Unset(key)
+}
+
+// UnsetGlobal delets a key from the global config.
+func (c *Configs) UnsetGlobal(key string) error {
+	if c.global == nil {
+		return nil
+	}
+
+	return c.global.Unset(key)
+}
+
+// Keys returns a list of all keys from all available scopes. Every key has section and possibly
+// a subsection. They are seprated by dots. The subsection itself may contain dots. The final
+// key name and the section MUST NOT contain dots.
+//
+// Examples
+//   - remote.gist.gopass.pw.path -> section: remote, subsection: gist.gopass.pw, key: path
+//   - core.timeout -> section: core, key: timeout
+func (c *Configs) Keys() []string {
+	keys := make([]string, 0, 128)
+
+	for _, cfg := range []*Config{
+		c.Preset,
+		c.system,
+		c.global,
+		c.local,
+		c.worktree,
+		c.env,
+	} {
+		if cfg == nil {
+			continue
+		}
+		if cfg.vars == nil {
+			continue
+		}
+		for k := range cfg.vars {
+			keys = append(keys, k)
+		}
+	}
+
+	return set.Sorted(keys)
+}
+
+// List returns all keys matching the given prefix. The prefix can be empty,
+// then this is identical to Keys().
+func (c *Configs) List(prefix string) []string {
+	return set.SortedFiltered(c.Keys(), func(k string) bool {
+		return strings.HasPrefix(k, prefix)
+	})
+}
+
+// ListSections returns a sorted list of all sections.
+func (c *Configs) ListSections() []string {
+	return set.Sorted(set.Apply(c.Keys(), func(k string) string {
+		section, _, _ := splitKey(k)
+
+		return section
+	}))
+}
+
+// ListSubsections returns a sorted list of all subsections
+// in the given section.
+func (c *Configs) ListSubsections(wantSection string) []string {
+	// apply extracts the subsection and matches it to the empty string
+	// if it doesn't belong to the section we're looking for. Then the
+	// filter func filters out any empty string.
+	return set.SortedFiltered(set.Apply(c.Keys(), func(k string) string {
+		section, subsection, _ := splitKey(k)
+		if section != wantSection {
+			return ""
+		}
+
+		return subsection
+	}), func(s string) bool {
+		return s != ""
+	})
```

Comment:
I guess this would not come up as a prime example of idiomatic Go xD 
I'll blame generics, but fair enough, it works too.

Reply:
Yeah, it could be rewritten to be a bit simpler ...